crates/svm-abi/layout/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+//! | 0 | 0 0 1 | 0 0 0 1 |  Amount - 2 bytes         |
+//! | 0 | 0 1 0 | 0 0 0 1 |  Amount - 3 bytes         |
+//! | 0 | 0 1 1 | 0 0 0 1 |  Amount - 4 bytes         |
+//! | 0 | 1 0 0 | 0 0 0 1 |  Amount - 5 bytes         |
+//! | 0 | 1 0 1 | 0 0 0 1 |  Amount - 6 bytes         |
+//! | 0 | 1 1 0 | 0 0 0 1 |  Amount - 7 bytes         |
+//! | 0 | 1 1 1 | 0 0 0 1 |  Amount - 8 bytes         |
+//! +---+-------+---------+---------------------------+
+//! | 0 | 0 0 0 | 0 0 1 0 |  i8  (signed)             |
+//! | 0 | 0 0 1 | 0 0 1 0 |  u8  (unsigned)           |
+//! | 0 | 0 1 0 | 0 0 1 0 |  i16 (signed)   - 1 byte  |
+//! | 0 | 0 1 1 | 0 0 1 0 |  i16 (signed)   - 2 bytes |
+//! | 0 | 1 0 0 | 0 0 1 0 |  u16 (unsigned) - 1 byte  |
+//! | 0 | 1 0 1 | 0 0 1 0 |  u16 (signed)   - 2 bytes |
+//! | 0 | 1 1 0 | 0 0 1 0 |  Reserved                 |
+//! | 0 | 1 1 1 | 0 0 1 0 |  Reserved                 |
+//! +---+-------+---------+---------------------------+
+//! | 0 | 0 0 0 | 0 0 1 1 |  i32 (signed)   - 1 byte  |
+//! | 0 | 0 0 1 | 0 0 1 1 |  i32 (signed)   - 2 bytes |
+//! | 0 | 0 1 0 | 0 0 1 1 |  i32 (signed)   - 3 bytes |
+//! | 0 | 0 1 1 | 0 0 1 1 |  i32 (signed)   - 4 bytes |
+//! | 0 | 1 0 0 | 0 0 1 1 |  i32 (unsigned) - 1 bytes |
+//! | 0 | 1 0 1 | 0 0 1 1 |  i32 (unsigned) - 2 bytes |
+//! | 0 | 1 1 0 | 0 0 1 1 |  i32 (unsigned) - 3 bytes |
+//! | 0 | 1 1 1 | 0 0 1 1 |  i32 (unsigned) - 4 bytes |
+//! +---+-------+---------+---------------------------+
+//! | 0 | 0 0 0 | 0 1 0 0 |  i64 (signed)   - 1 byte  |
+//! | 0 | 0 0 1 | 0 1 0 0 |  i64 (signed)   - 2 bytes |
+//! | 0 | 0 1 0 | 0 1 0 0 |  i64 (signed)   - 3 bytes |
+//! | 0 | 0 1 1 | 0 1 0 0 |  i64 (signed)   - 4 bytes |
+//! | 0 | 1 0 0 | 0 1 0 0 |  i64 (signed)   - 5 bytes |
+//! | 0 | 1 0 1 | 0 1 0 0 |  i64 (signed)   - 6 bytes |
+//! | 0 | 1 1 0 | 0 1 0 0 |  i64 (signed)   - 7 bytes |
+//! | 0 | 1 1 1 | 0 1 0 0 |  i64 (signed)   - 8 bytes |
+//! +---+-------+---------+---------------------------+
+//! | 0 | 0 0 0 | 0 1 0 1 |  i64 (unsigned) - 1 byte  |
+//! | 0 | 0 0 1 | 0 1 0 1 |  i64 (unsigned) - 2 bytes |
+//! | 0 | 0 1 0 | 0 1 0 1 |  i64 (unsigned) - 3 bytes |
+//! | 0 | 0 1 1 | 0 1 0 1 |  i64 (unsigned) - 4 bytes |
+//! | 0 | 1 0 0 | 0 1 0 1 |  i64 (unsigned) - 5 bytes |
+//! | 0 | 1 0 1 | 0 1 0 1 |  i64 (unsigned) - 6 bytes |
+//! | 0 | 1 1 0 | 0 1 0 1 |  i64 (unsigned) - 7 bytes |
+//! | 0 | 1 1 1 | 0 1 0 1 |  i64 (unsigned) - 8 bytes |
+//! +---+-------+---------+---------------------------+
+//! | 0 | 0 0 0 | 0 1 1 0 |  Array - 0 items          |
+//! | 0 | 0 0 1 | 0 1 1 0 |  Array - 1 item           |
+//! | 0 | 0 1 0 | 0 1 1 0 |  Array - 2 items          |
+//! | 0 | 0 1 1 | 0 1 1 0 |  Array - 3 items          |
+//! | 0 | 1 0 0 | 0 1 1 0 |  Array - 4 items          |
+//! | 0 | 1 0 1 | 0 1 1 0 |  Array - 5 items          |
+//! | 0 | 1 1 0 | 0 1 1 0 |  Array - 6 items          |
+//! | 0 | 1 1 1 | 0 1 1 0 |  Array - 0..255 items     |
```

Comment:
Are you keeping the leading bit for something else, or was it just that there was no need to use it? You always separate him from the next 3 bits, so I was wandering.

Reply:
yes, the MSB bit stands for `has more`

Reply:
that way we could always had more types.
However, I don't expect that we'll need more than a single byte as it's now